packages/framework/src/Component/Translation/ConstraintViolationExtractor.php

```diff
@@ -0,0 +1,188 @@
+<?php
+
+namespace Shopsys\FrameworkBundle\Component\Translation;
+
+use JMS\TranslationBundle\Model\FileSource;
+use JMS\TranslationBundle\Model\Message;
+use JMS\TranslationBundle\Model\MessageCatalogue;
+use JMS\TranslationBundle\Translation\Extractor\FileVisitorInterface;
+use PhpParser\Node;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Expr\Variable;
+use PhpParser\Node\Name\FullyQualified;
+use PhpParser\Node\Scalar\String_;
+use PhpParser\Node\Stmt\ClassMethod;
+use PhpParser\NodeTraverser;
+use PhpParser\NodeVisitor;
+use PhpParser\NodeVisitor\NameResolver;
+use SplFileInfo;
+use Symfony\Component\Validator\Context\ExecutionContextInterface;
+use Twig_Node;
+
+/**
+ * Extracts custom message from constraint callback function.
+ *
+ * Examples:
+ *     Working example
+ *     public function workingExample(ExecutionContextInterface $context)
+ *      {
+ *          $context->addViolation('This message will be extracted into "validators" translation domain');
+ *      }
+ *
+ *      non-working example
+ *      public function nonWorkingExample(ExecutionContextInterface $context)
+ *      {
+ *          $message = 'This message will be not extracted into "validators" translation domain';
+ *          $context->addViolation($message);
+ *      }
+ */
+class ConstraintViolationExtractor implements FileVisitorInterface, NodeVisitor
+{
+    /**
+     * @var \PhpParser\NodeTraverser
+     */
+    private $traverser;
+
+    /**
+     * @var \JMS\TranslationBundle\Model\MessageCatalogue
+     */
+    private $catalogue;
+
+    /**
+     * @var \SplFileInfo
+     */
+    private $file;
+
+    /**
+     * @var string[]
+     */
+    private $currentExecutionContextVariableNames;
+
+    public function __construct()
+    {
+        $this->traverser = new NodeTraverser();
+        $this->traverser->addVisitor(new NameResolver());
+        $this->traverser->addVisitor($this);
+        $this->currentExecutionContextVariableNames = [];
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function visitPhpFile(SplFileInfo $file, MessageCatalogue $catalogue, array $ast)
+    {
+        $this->file = $file;
+        $this->catalogue = $catalogue;
+        $this->traverser->traverse($ast);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function enterNode(Node $node)
+    {
+        $this->setCurrentExecutionContextVariableNamesFromNode($node);
+        if ($node instanceof MethodCall && $this->isAddViolationMethodCall($node)) {
+            $this->extractMessage($node);
+        }
+    }
+
+    /**
+     * @param \PhpParser\Node $node
+     */
+    private function setCurrentExecutionContextVariableNamesFromNode(Node $node)
```

Comment:
This method does not do what it says it does - it does nothing in majority of cases, and it only adds new variable names.

I would suggest moving the `instanceof` condition to the `enterNode` method, which would provide context to someone who reads the method - he would see in which nodes some context is changes. Also, it would let you typehint the parameter of `setCurrentExecutionContextVariableNamesFromNode` as `ClassMethod`, improving the IDE autocompletion.

Also, at the beginning of this method's body, the `$currentExecutionContextVariableNames` should be reset to an empty array, otherwise it would only add new variable names and it would depend on reseting the property by another method.

Suggested changes:
```php
    public function enterNode(Node $node)
    {
        if ($node instanceof ClassMethod) {
            $this->setCurrentExecutionContextVariableNamesFromNode($node);
        } elseif ($node instanceof MethodCall && $this->isAddViolationMethodCall($node)) {
            $this->extractMessage($node);
        }
    }

    private function setCurrentExecutionContextVariableNamesFromNode(ClassMethod $node)
    {
        $this->currentExecutionContextVariableNames = [];
        foreach ($node->getParams() as $parameter) {
            if ($this->isParameterExecutionContextInterfaceSubclass($parameter)) {
                $this->currentExecutionContextVariableNames[] = $parameter->name;
            }
        }
    }
```

Reply:
Hi @sspooky13, could you please look at this comment and implement the requested change? :slightly_smiling_face: